### ISSUE TITLE
Define CSL-POSE (or CSL-LOSE?)

### ISSUE THREAD

**Issue by omasanori** [omasanori/schemebib#4]
[CSL-JSON](https://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html) ([schema](https://github.com/citation-style-language/schema)) but [POSE](https://www.pose.s-expressions.org/abstract) or [LOSE](https://srfi.schemers.org/srfi-176/srfi-176.html#_part_2_line_oriented_s_expressions_lose) would be great for us.

**Comment by omasanori** [comment]
First thought:
```lisp
; @InProceedings{BaC:2011a,
;   author =       "Eli Barzilay and Ryan Culpepper and Matthew Flatt",
;   title =        "{Keeping it Clean with Syntax Parameters}",
;   booktitle =    proc-Scheme,
;   year =         "2011",
;   series =       "Scheme",
;   URL =          "http://www.schemeworkshop.org/2011/papers/Barzilay2011.pdf",
;   abstract =     "Racket's syntax parameters support the hygienic
;                   implementation of syntactic forms that would otherwise
;                   introduce implicit identifiers unhygienically.",
; }

(item (id BaC-2011a)
      (type paper-conference)
      (author ((family "Barzilay") (given "Eli"))
              ((family "Ryan") (given "Culpepper"))
              ((family "Matthew") (given "Flatt")))
      (title "Keeping it Clean with Syntax Parameters")
      (container-title "Proceedings of the Workshop on Scheme and Functional Programming")
      (issued "2011-10-23")
      (collection-title "Scheme")
      (URL "http://www.schemeworkshop.org/2011/papers/Barzilay2011.pdf")
      (abstract "Racket's syntax parameters support the hygienic implementation of syntactic forms that would otherwise introduce implicit identifiers unhygienically."))
```

**Comment by omasanori** [comment]
```json
[{"id": "BaC:2011a",
  "type": "paper-conference",
  "author": [{"given": "Eli", "family": "Barzilay"},
             {"given": "Ryan", "family": "Culpepper"},
             {"given": "Matthew", "family": "Flatt"}],
  "title": "Keeping it Clean with Syntax Parameters",
  "container-title": "Proceedings of the Workshop on Scheme and Functional Programming",
  "issued": "2011",
  "collection-title": "Scheme",
  "URL": "http://www.schemeworkshop.org/2011/papers/Barzilay2011.pdf",
  "abstract": "Racket's syntax parameters support the hygienic implementation of syntactic forms that would otherwise introduce implicit identifiers unhygienically."}]
```
So far, so good.

**Comment by omasanori** [comment]
My current workflow is:

1. Compile bibliographic information into `schemebib.yaml` as a *CSL-YAML* file
2. Convert it to a CSL-JSON file
3. Check it with LuaLaTeX using `citation-style-language` CTAN package

**Comment by lassik** [comment]
Thanks for looking into this.

I'm finding the CSL "ecosystem" hard to understand as an outsider.

* CSL stands for Citation **Style** Language. They have a [repo with thousands of styles](https://github.com/citation-style-language/styles) described in Lisp-like XML. There is no obvious material dealing with bibliography entries in the CSL spec. It only deals with styling.

* The schema repo to which you link has two JSON schemas: "CSL input data" and "CSL citation objects". It says these are unofficial extensions to base CSL, but widely used. Base CSL is XML, but these are JSON. People are also using YAML.

BibTeX and CSL interop exists, but [seems to be non-trivial](https://forums.zotero.org/discussion/6375/idea-using-csl-citation-styles-to-output-biblio-in-latex).

As a separate project from CSL, there's also [CFF: Citation File Format](https://citation-file-format.github.io/). This is used for the "cite this repo" feature of GitHub - see e.g. the [Ribbit repo](https://github.com/udem-dlteam/ribbit).

**Comment by lassik** [comment]
I would summarize the relevant facts as follows.

* BibTeX isn't going away any time soon.
* CSL has a complicated ecosystem. People are mixing various standard and non-standard parts, various tools and formats. All of those are still evolving.
* CFF is different from CSL. There is no obvious plan to join them.

Given that the situation is so messy, my recommendation is that we use a custom S-expression schema that contains the data we need. We are free to add as much data as is required to achieve good conversions into BibTeX, CSL, and CFF. Where possible, we should re-use parts of the schemas of these formats in our S-expression schema instead.

**Comment by lassik** [comment]
I suggest we use the following workflow:

1. Start collecting some S-expression bibliography entries.
2. Write scripts that convert those to BibTeX, CSL JSON, an CFF YAML.
3. Add extra information to the S-expression entries until the scripts achieve a high-quality conversion to each target format.
4. Collect more bibliography entries, repeat.

The existing `.scm` files in [schemedoc/bibliography](https://github.com/schemedoc/bibliography) use LOSE (line-oriented S-expressions), i.e. each top-level S-expression on its own line, similar to [JSON Lines](https://jsonlines.org/). If this is too limiting, I don't mind switching to the more general POSE.

If we are using S-expressions, it's also easy to write macros to reduce repetition in the source data as needed.

**Comment by lassik** [comment]
> Compile bibliographic information into schemebib.yaml as a CSL-YAML file

Are you getting the information from schemedoc/bibliography or from somewhere else?

**Comment by omasanori** [comment]
Thank you for your survey!

I found the bibliography format draft by the way: https://github.com/s-expressions/dsl

**Comment by omasanori** [comment]
> Are you getting the information from schemedoc/bibliography or from somewhere else?

The primary source is the archive of readscheme.org like schemedoc/bibliography. In the future, we will collect things newer than 2011.

As readscheme.org contains some articles not strictly related to Scheme but relevant to Lisp and ML families, I take the phrase *Scheme-related* in a broad sense. I think it is fine to include all articles in certain series, ex. ELS, ILC, LFP, Lisp Pointers as well as Scheme Workshop.

**Comment by lassik** [comment]
> In the future, we will collect things newer than 2011.

We've been collecting them here: https://github.com/schemedoc/bibliography/issues/6

> As readscheme.org contains some articles not strictly related to Scheme but relevant to Lisp and ML families, I take the phrase _Scheme-related_ in a broad sense. I think it is fine to include all articles in certain series, ex. ELS, ILC, LFP, Lisp Pointers as well as Scheme Workshop.

Agreed. There should be a broader "Lisp bibliography", but I'm not aware of any attempts to make one, so we can start doing it.

If we use our own schema, we can add tags to each paper. That makes it easy to filter the collection, e.g. only take the papers with "Scheme" tag.

**Comment by lassik** [comment]
@omasanori Are you in general agreement with the above plan? How do you want to split the work?

**Comment by lassik** [comment]
When I worked on schemedoc/bibliography I started by converting each `page<N>.md` into an equivalent `page<N>.scm`, containing the same papers in the same order. IMHO this is the cleanest way to do it. Then we get a 1:1 conversion of the old ReadScheme before we add new things to it. This makes it easiest to verify that we didn't accidentally drop any papers.

**Comment by omasanori** [comment]
In principle I agree with you. This repo will be kinda sandbox to be merged into the schemedoc one.
Regarding file organization, I prefer single file with tagging items or file-per-item over categories as some items belong to multiple categories so duplication is unavoidable with categories. On the other hand, I agree with you on the verification point of view. Let's keep the status quo and think dedup later.

**Comment by lassik** [comment]
> This repo will be kinda sandbox to be merged into the schemedoc one.

Fair enough. Are you using Scheme to write the tools (converters, etc.)?

> Regarding file organization, I prefer single file with tagging items or file-per-item over categories as some items belong to multiple categories so duplication is unavoidable with categories.

I agree that file-per-category is problematic.

My preference (after ReadScheme has been converted) would be one file per year, containing all papers from that year. This means the older files will eventually stop changing. More stability is nice.

Putting all papers in one file will result in a huge file that is harder to edit.

**Comment by omasanori** [comment]
Yes, per-year scheme sounds reasonable.

> Are you using Scheme to write the tools (converters, etc.)?

Not yet (I am using existing tools written in Perl, Rust and such), but from now on, as we will work together with an S-expression based format.

**Comment by lassik** [comment]
Sounds good. I'll wait a while until you have a prototype you are happy with. Let me know if you need help with something.

**Comment by omasanori** [comment]
Current idea: use the bib DSL (POSE), using BibLaTeX types and field names (BibTeX has less precision and we will use `misc` too much for irregular entries, and BibLaTeX standard data model is documented better than CSL-JSON data model) with CFF and CSL-JSON types and field names where BibLaTeX does not model well. MARC, BIBFRAME, DC, Schema.org, Linked Data etc: they are indeed interesting but let's revisit later.

**Comment by omasanori** [comment]
Well, I will start with LOSE, as if it works to me then we can continue with existing losebib tools.

**Comment by lassik** [comment]
Sounds excellent. Thanks for putting effort into this.

**Comment by lassik** [comment]
BTW, [Falsehoods Programmers Believe About Names](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/) may come in handy.

I noticed something weird in the CSL-JSON schema, where they stuff all unknown name parts into the "surname" field. Not sure if BibTeX, BibLaTeX, and CFF have similar flaws.

**Comment by omasanori** [comment]
> BTW, [Falsehoods Programmers Believe About Names](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/) may come in handy.

Yeah, I really like that one. [Text Rendering Hates You](https://faultlore.com/blah/text-hates-you/) and [Text Editing Hates You Too](https://lord.io/text-editing-hates-you-too/) are also my favorite ones. I am wondering if someone has published similar one for calendars...

Anyways, BibTeX (and BibLaTeX's BibTeX-compatible format) is more naive "Last name, first name" model if I understand correctly. BibLaTeX and CFF support more *structured* name format similar to CSL-JSON. Something similar to the [personal name](https://github.com/s-expressions/dsl/blob/b7af195a58a28397897513db101b29f150a3a00a/spec/personal-name.md) spec's "full name and sort key" model is also supported by BibLaTeX (`sortname`).

**Comment by lassik** [comment]
> Yeah, I really like that one. [Text Rendering Hates You](https://faultlore.com/blah/text-hates-you/) and [Text Editing Hates You Too](https://lord.io/text-editing-hates-you-too/) are also my favorite ones.

Nice.

> I am wondering if someone has published similar one for calendars...

[Yes, several.](https://github.com/kdeldycke/awesome-falsehood#dates-and-time)

> Anyways, BibTeX (and BibLaTeX's BibTeX-compatible format) is more naive "Last name, first name" model if I understand correctly. BibLaTeX and CFF support more _structured_ name format similar to CSL-JSON. Something similar to the [personal name](https://github.com/s-expressions/dsl/blob/b7af195a58a28397897513db101b29f150a3a00a/spec/personal-name.md) spec's "full name and sort key" model is also supported by BibLaTeX (`sortname`).

Right. We probably copied the sortname idea from BibLaTeX.

I don't know what the right amount of structure is. Whatever produces a good conversion, perhaps.

**Comment by omasanori** [comment]
We have instances of name with suffix that are absolutely unavoidable: Norman I. Adams *IV* and Guy L. Steele *Jr.*

We need to generate something like `Steele Jr., Guy L.` or `Guy L. {Steele Jr.}` for BibTeX-ish tools. Otherwise `Jr.` will be treated as the last name in the worst case. Thus, full name + sort key is excellent but we can not use this one exclusively.

If we choose two-component model as the second option, `Guy L.` is first name and `Steele Jr.` is last name. If we want to avoid such artificial separation, then we will implement the all four or five components BibLaTeX, CSL-JSON and CFF provide. (and L. will be part of first name anyways since implementing more complex model won't benefit us as John Cowan noted somewhere)

In short, we have two options: BibTeX model or BibLaTeX model.

**Comment by lassik** [comment]
Maybe we should have all of the components that the others have? What do you think?

**Comment by omasanori** [comment]
We will not do sorting itself but storing components separately, so having many components does not increase complexity of our code significantly. So, yes, it would be fine to have all components.

**Comment by lassik** [comment]
Sounds good. If we need to sort by author, we can add a sort key field later.

**Comment by omasanori** [comment]
OK, as a first step, I will send a short specification of LoseBib describing the status quo to the schemedoc one.
Then, I will write proposals of extensions as PRs updating the spec. After a extension spec PR is merged, a PR applying the extension to existing entries.
By doing so new comers will get involved easier.

**Comment by lassik** [comment]
Good plan.

Unfortunately it's not easy to persuade newcomers to work together with other schemers on the same projects, but we can hope :)

**Comment by omasanori** [comment]
I made https://github.com/schemedoc/bibliography/pull/16 and we will continue on the schemedoc repo.
